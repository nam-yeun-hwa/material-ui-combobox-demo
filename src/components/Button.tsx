type Props = {
  children: React.ReactNode;
  activeClass?: string;
  onClickHandler: () => void;
  dataTestId: string;
};

export default function Button({
  children,
  activeClass,
  onClickHandler,
  dataTestId,
}: Props) {
  return (
    <>
      <button
        data-testid={dataTestId}
        className={activeClass}
        onClick={onClickHandler}
        onMouseDown={(e) => {
          e.preventDefault();
        }}
      >
        {children}
      </button>
    </>
  );
}
