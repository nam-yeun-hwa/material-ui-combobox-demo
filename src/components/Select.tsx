import { ChangeEvent, KeyboardEvent, useEffect, useRef, useState } from "react";
import { optionItem, selectOptionType } from "type/data";
import "./select.css";
import OptionItem from "./OptionItem";
import useWindowSize from "../hook/useWindowSize.ts";
import { SELECT_OPTION } from "../constant/constant";
import Button from "./Button.tsx";

type SelectProps = {
  options: selectOptionType | undefined;
};

export default function Select(props: SelectProps) {
  const { options } = props;
  const [isFocused, setIsFocused] = useState(false);

  const [inputValue, setInputValue] = useState("");
  const [optionSearchList, setOptionSearchList] = useState<
    selectOptionType | undefined
  >();
  const [selectMaxWidth, setSelectMaxWidth] = useState(300);

  const [selectOptionActive, setSelectOptionActive] = useState<
    optionItem | undefined
  >();

  const [enterItem, setOnEnterItem] = useState<optionItem | undefined>();
  const [isOptionToggle, setIsOptionToggle] = useState(false);

  const optionRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);
  const selectRef = useRef<HTMLDivElement>(null);

  const windowSize = useWindowSize();

  const [availableSpace, setAvailableSpace] = useState<{
    top: Number | undefined;
    bottom: Number | undefined;
  }>();

  /**
   * @description select 엘레멘트 Y값 가져오기
   */
  useEffect(() => {
    if (selectRef.current) {
      // 셀렉트 옵션 하단 넓이
      const bottom_height =
        windowSize.height -
        selectRef.current.getBoundingClientRect().top -
        SELECT_OPTION.HEIGHT;

      //셀렉트 옵션 상단 넓이
      const top_height =
        windowSize.height - bottom_height - SELECT_OPTION.HEIGHT;
      setAvailableSpace({ top: top_height, bottom: bottom_height });
    }
  }, [selectRef, windowSize]);

  /**
   * @description select component 최대값 구하기
   */
  useEffect(() => {
    if (optionRef.current) {
      const width = optionRef.current.offsetWidth;
      setSelectMaxWidth((preState) => {
        return Math.max(preState, width);
      });
    }
  }, [options, selectMaxWidth]);

  /**
   * @description input검색에 따른 option filter
   */
  useEffect(() => {
    if (inputValue.length > 0) {
      const filteredItems = options?.filter((optionItem) => {
        return optionItem.label.includes(inputValue);
      });

      if (filteredItems && filteredItems?.length > 0) {
        setOptionSearchList(filteredItems);
      } else {
        setOptionSearchList([{ label: "No options", value: "0" }]);
      }
    }
  }, [inputValue, setOptionSearchList]);

  /**
   * @description selectOptionActive 초기화 시켜주기
   * option 클릭 후 input 해당 옵션값의 텍스트가 들어가 있을때 옵션 값을 다 지워줄 경우에는
   * selectOptionActive를 false로 초기화 해준다.
   */
  useEffect(() => {
    if (inputValue.length === 0) {
      setSelectOptionActive(undefined);
    }
  }, [inputValue]);

  /**
   * @function handleInputChange
   * @description select option 검색
   */
  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setInputValue(event.target.value);
  };

  /**
   * @function onFocusHandler
   * @description 포커스 이벤트 핸들러
   */
  const onFocusHandler = () => {
    setIsFocused(true);
    setIsOptionToggle(true);
  };

  /**
   * @function onBlurHandler
   * @deprecated 포커스 아웃 이벤트 핸들러
   */
  const onBlurHandler = () => {
    setIsFocused(false);
    setInputValue("");
  };

  /**
   * @function onClickOptionValue
   * @param optionValue 클릭한 option item
   * @description 클릭한 옵션
   */
  const onClickOptionItem = (optionItem: optionItem) => {
    setIsOptionToggle(false);
    if (optionItem) {
      setInputValue(optionItem.label);
      setSelectOptionActive(optionItem);
    }
  };

  /**
   * @function onClickClearInputSearch
   * @description input search clear
   */
  const onClickClearInputSearch = () => {
    setInputValue("");
    setIsFocused(true);
    setOnEnterItem(undefined);
    setSelectOptionActive(undefined);
    if (inputRef.current) {
      inputRef.current.focus();
    }
  };

  /**
   * @function onClickFocusOnHandler
   * @description input 클릭시 Focus On
   */
  const onClickFocusOnHandler = () => {
    setIsFocused(true);
    setIsOptionToggle(true);
  };

  /**
   * @function onToggleOptionList
   * @description option toggle
   */
  const onToggleOptionList = () => {
    // setIsFocused(!isFocused);
    setIsOptionToggle(!isOptionToggle);
  };

  /**
   * @function onMouseOverHandler
   * @param optionItem
   */
  const onMouseOverHandler = (optionItem: optionItem) => {
    setOnEnterItem(optionItem);
  };

  /**
   * @function onKeyboardHandler
   * @description 키보드 이벤트
   */
  const onKeyboardHandler = (e: KeyboardEvent<HTMLDivElement>) => {
    switch (e.key) {
      case "Enter":
        console.log("Enter 키가 눌렸습니다.");
        if (enterItem) {
          onClickOptionItem(enterItem);
        }
        break;
      case "ArrowUp":
        console.log("위쪽 화살표 키가 눌렸습니다.");
        if (isOptionToggle) {
          setOnEnterItem((prevState) => {
            if (prevState) {
              const currentValue = Number(prevState?.value);
              return options?.find(
                (item) => item.value === String(currentValue - 1)
              );
            } else {
              return options && options[99];
            }
          });
        } else {
          setIsOptionToggle(true);
        }

        break;
      case "ArrowDown":
        if (isOptionToggle) {
          setOnEnterItem((prevState) => {
            if (prevState) {
              const currentValue = Number(prevState?.value);
              return options?.find(
                (item) => item.value === String(currentValue + 1)
              );
            } else {
              return options && options[0];
            }
          });
        } else {
          setIsOptionToggle(true);
        }
        console.log("아래쪽 화살표 키가 눌렸습니다.");

        break;
      default:
        break;
    }
  };

  return (
    <>
      <div
        className={`select-component`}
        ref={selectRef}
        data-testid="select-component"
      >
        <div
          className={`MuiAutocomplete-root`}
          style={{ width: `${selectMaxWidth}px` }}
          onKeyDown={onKeyboardHandler}
          tabIndex={0}
        >
          <div className="MuiFormControl-root">
            {/* <label className="MuiFormLabel-root">Movie</label> */}
            <div
              className={`MuiInputBase-root ${isFocused && "focused"} ${
                selectOptionActive && "clearVisible"
              }`}
            >
              <input
                className="MuiInputBase-input"
                ref={inputRef}
                type="text"
                value={inputValue}
                onClick={onClickFocusOnHandler}
                onFocus={onFocusHandler}
                onBlur={onBlurHandler}
                onChange={handleInputChange}
                placeholder={"Movie"}
              />
              <div className="MuiAutocomplete-endAdornment">
                <Button
                  dataTestId={"clear"}
                  activeClass={`MuiAutocomplete-clearIndicator ${
                    selectOptionActive &&
                    inputValue.length > 0 &&
                    "clearIndicator-visible"
                  }`}
                  children={
                    <svg>
                      <path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path>
                    </svg>
                  }
                  onClickHandler={onClickClearInputSearch}
                />

                <Button
                  dataTestId={"toggle"}
                  activeClass="MuiAutocomplete-arrowIndicator"
                  children={
                    <svg>
                      <path d="M7 10l5 5 5-5z"></path>
                    </svg>
                  }
                  onClickHandler={onToggleOptionList}
                />
              </div>
            </div>
          </div>
        </div>
        <div
          data-testid="optionToggle"
          className={`base-popper-root ${
            isFocused && isOptionToggle && "open"
          } ${
            availableSpace?.top &&
            availableSpace?.bottom &&
            availableSpace.top > availableSpace.bottom &&
            "base-popper-root_bottom"
          }`}
          style={{ width: `${selectMaxWidth > 300 && selectMaxWidth}px` }}
        >
          <div className="base-popper-content" ref={optionRef}>
            {/* 1. 선택항목이 없고 input값이 없을때 > 오리지날 option list*/}
            {selectOptionActive === undefined && inputValue?.length === 0 && (
              <OptionItem
                options={options}
                onClickOptionItem={onClickOptionItem}
                onMouseOverHandler={onMouseOverHandler}
                activeItem={selectOptionActive}
                hoverItem={enterItem}
              />
            )}
            {/* 2. 선택항목이 없고 검색 input이 있을때 > 검색리스트 */}
            {selectOptionActive === undefined &&
              inputValue &&
              inputValue?.length > 0 && (
                <OptionItem
                  options={optionSearchList}
                  onClickOptionItem={onClickOptionItem}
                  onMouseOverHandler={onMouseOverHandler}
                  activeItem={selectOptionActive}
                  hoverItem={enterItem}
                />
              )}

            {/* 3. 선택항목이 있을때 > 선택항목이 Active 된 option list */}
            {selectOptionActive && (
              <OptionItem
                options={options}
                onClickOptionItem={onClickOptionItem}
                onMouseOverHandler={onMouseOverHandler}
                activeItem={selectOptionActive}
                hoverItem={enterItem}
              />
            )}
            {/* 4. focus상태이고 선택된 항목이 있을때 */}
            {isOptionToggle && selectOptionActive && (
              <OptionItem
                options={options}
                onClickOptionItem={onClickOptionItem}
                onMouseOverHandler={onMouseOverHandler}
                activeItem={selectOptionActive}
                hoverItem={enterItem}
              />
            )}
          </div>
        </div>
      </div>
    </>
  );
}
