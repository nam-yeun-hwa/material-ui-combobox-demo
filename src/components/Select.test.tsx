import { fireEvent, render, screen } from "@testing-library/react";
import Select from "./Select";
import Button from "./Button";

const options = [
  { value: "1", label: "Option 1" },
  { value: "2", label: "Option 2" },
  { value: "3", label: "Option 3" },
];

describe("<Select />", () => {
  test("Select 컴포넌트가 렌더링되는지 확인", () => {
    const { container } = render(<Select options={options} />);
    const selectComponent = screen.getByTestId("select-component");
    expect(selectComponent).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("input이 올바르게 렌더링되는지 확인", () => {
    render(<Select options={options} />);
    const inputElement = screen.getByPlaceholderText("Movie");
    expect(inputElement).toBeInTheDocument();
  });

  test("input 입력값에 따라 옵션 리스트가 필터링되는지 확인", () => {
    render(<Select options={options} />);
    const inputElement = screen.getByPlaceholderText("Movie");
    fireEvent.change(inputElement, { target: { value: "Option 1" } });
    const filteredOption = screen.getByText("Option 1");
    expect(filteredOption).toBeInTheDocument();
  });

  test("input change 이벤트가 올바르게 동작하는지 확인", () => {
    render(<Select options={options} />);
    const inputElement = screen.getByPlaceholderText(
      "Movie"
    ) as HTMLInputElement;
    fireEvent.change(inputElement, { target: { value: "Lord" } });
    expect(inputElement.value).toBe("Lord");
  });

  test("클리어 버튼 클릭 시 input에 내용이 지워지는지 확인", () => {
    render(<Select options={options} />);
    const inputElement = screen.getByPlaceholderText(
      "Movie"
    ) as HTMLInputElement;
    fireEvent.change(inputElement, { target: { value: "Lord" } });
    expect(inputElement.value).toBe("Lord");

    const onClickHandler = jest.fn();
    render(
      <Button dataTestId="" onClickHandler={onClickHandler}>
        Click Me
      </Button>
    );

    const clearBtn = screen.getByTestId("clear");
    // clear 버튼 클릭
    fireEvent.click(clearBtn);
    // 값이 비워졌는지 확인
    expect(inputElement.value).toBe("");
  });

  test("셀렉트 컴포넌트 토글버튼 클릭 이벤트가 잘 동작하는지 확인", () => {
    const handlerClick = jest.fn();
    const { getByTestId } = render(
      <Button dataTestId="toggle" onClickHandler={handlerClick}>
        toggle button
      </Button>
    );
    expect(handlerClick).toHaveBeenCalledTimes(0);
    const toggleBtn = getByTestId("toggle");
    fireEvent.click(toggleBtn);

    expect(handlerClick).toHaveBeenCalledTimes(1);
  });

  it("onKeyboardHandler 잘 동작하는지 확인", () => {
    const onKeyboardHandler = jest.fn();
    const { container } = render(<Select options={options} />);

    const divElement = container.querySelector(
      ".MuiAutocomplete-root"
    ) as HTMLDivElement;

    fireEvent.keyDown(divElement, { key: "Enter" });
    expect(onKeyboardHandler).toHaveBeenCalledTimes(0);

    fireEvent.keyDown(divElement, { key: "ArrowUp" });
    expect(onKeyboardHandler).toHaveBeenCalledTimes(0);

    fireEvent.keyDown(divElement, { key: "ArrowDown" });
    expect(onKeyboardHandler).toHaveBeenCalledTimes(0);
  });
});
