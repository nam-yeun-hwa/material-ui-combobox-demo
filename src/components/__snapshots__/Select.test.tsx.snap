// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> Select 컴포넌트가 렌더링되는지 확인 1`] = `
<div>
  <div
    class="select-component"
    data-testid="select-component"
  >
    <div
      class="MuiAutocomplete-root"
      style="width: 300px;"
      tabindex="0"
    >
      <div
        class="MuiFormControl-root"
      >
        <div
          class="MuiInputBase-root false undefined"
        >
          <input
            class="MuiInputBase-input"
            placeholder="Movie"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment"
          >
            <button
              class="MuiAutocomplete-clearIndicator undefined"
              data-testid="clear"
            >
              <svg>
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <button
              class="MuiAutocomplete-arrowIndicator"
              data-testid="toggle"
            >
              <svg>
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="base-popper-root false 0"
      data-testid="optionToggle"
    >
      <div
        class="base-popper-content"
      >
        <div
          class="option-item false"
          data-testid="option-list"
        >
          Option 1
        </div>
        <div
          class="option-item false"
          data-testid="option-list"
        >
          Option 2
        </div>
        <div
          class="option-item false"
          data-testid="option-list"
        >
          Option 3
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select /> renders component correctly 1`] = `
<div>
  <div
    class="select-component"
  >
    <div
      class="MuiAutocomplete-root"
      style="width: 300px;"
      tabindex="0"
    >
      <div
        class="MuiFormControl-root"
      >
        <div
          class="MuiInputBase-root false undefined"
        >
          <input
            class="MuiInputBase-input"
            placeholder="Movie"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment"
          >
            <button
              class="MuiAutocomplete-clearIndicator undefined"
            >
              <svg>
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <button
              class="MuiAutocomplete-arrowIndicator"
            >
              <svg>
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="base-popper-root false 0"
    >
      <div
        class="base-popper-content"
      >
        <div
          class="option-item false"
        >
          Option 1
        </div>
        <div
          class="option-item false"
        >
          Option 2
        </div>
        <div
          class="option-item false"
        >
          Option 3
        </div>
      </div>
    </div>
  </div>
</div>
`;
